### PR TITLE
Fix bug when switching window content in winform

### PR DESCRIPTION
**Pull request by saroad2**
Bug Description
==========

In my application I'm using a single `toga.Window` and switch its content multiple times instead of opening new windows.
In my main box `box1` I have button that replace the content of the window with `box2`, and in `box2` I have a "back" button that replace the content of the window back to `box1`.

When I'm clicking the button that move from `box1` to `box2` everything is fine. When I try to got back to to `box1` using the "back" button I get a `RuntimeError` that "Already have a container".

This is a bug.

How I fixed it
=========
When setting the container of a widget in windows, first if the widget already has a container, setting it to None and removing it from `Controls`. After that, if the new container is not None, setting it.

This way we can remove boxes from the window and move them back freely.

## PR Checklist:

- [X] All new features have been tested
- [X] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

### CODE REVIEW

**Comment by mhsmith** [comment]
Related:
* #1509
* #1510

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
So - I'm unclear if this is a deliberate or unintentional change in behavior.

The "raise an exception if you assign a widget to a container that is already assigned to a container" behavior was intentional. It was intended to protect against accidentally "shifting" widgets. The intended behavior was that you would remove a widget from it's existing container, then assign it to a new container.

If this was an intentional change (based on your PR description, I suspect it might be?), then this PR by itself results in inconsistent behavior between backends - we need the analogous change on every other backend. However, before we do that, we should have a larger conversation about the API for container management, and whether this change is actually desirable at all. Do you have an argument for why this case should relocate the widget in a layout without requiring an explicit removal first? 

If it was an unintentional change, then there's obviously a gap in the implementation, as the exception will no longer be raised.

Either way, the other major requirement for this PR is a test case. The only example of widget "reparenting" in the examples at present is in the layout example, and that's not at the level of a window.  This is clearly an example we need - if only for testing this change - but also because this keeps coming up (as @mhsmith has flagged with related discussions).

**Comment by saroad2** [comment]
Hey @freakboy3742 , thanks for the elaborated response!

First of all, as you requested, I added a new example app called "One Window" that has one window with changing content box. This example works after my fix, but if you revert my initial commit it breaks when you try to get back to the "welcome box".

As for the removal of the `RuntimeError`, this was intentional. I don't see any reason to raise this since we do want sometimes to change the container of a box from one window to another.

The way I see it, there are 3 possible solutions to the bug as it is presented here:

Option 1 - Widget level - Remove old container and than set new one
-----------------------------------------------------------------------------

That was the approach I took. When setting a new container for a widget `widget1`, first I remove the old one (if it exists), than set the new one (if it's not None) and finally do the same for the widget children

Option 2 - Window implementation level - First set container to None, than set new container
--------------------------------------------------------------------------------------------------

In the window implementation (platform-wise), first set the container of the widget to None, causing the old container to be removed from the widget and its children, than set window as the new container.

Option 3 - Window interface level - Set container of old content to None
--------------------------------------------------------------------------------

In the `Window` interface, when switching content (as in this example) from `box1` to `box2`, first set the container of `box1` to None, than set this window as the container of `box2`



Those are the solutions I thought about. Maybe you have other ones.
Let me know what the solution you prefer, and I'll implement it.

**Comment by saroad2** [comment]
@mhsmith , I didn't know you were already working on it, so I'm sorry for the work duplication 🙏

**Comment by mhsmith** [comment]
No problem, I didn't have time to pursue that PR, so I've closed it in favor of this one.

The main difference between the PRs is the way they handle removing the native widget. I found that removing all controls in a loop also removed the menu bar: are you sure that `self.native.Controls.Remove` doesn't have the same problem?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
So - I think the approach you've taken here is the right one - this is a widget specific issue, so that's where the fix should lie. There is definitely an overlap/edge case when dealing with window content, but that doesn't mean the implementation for container handling is a window-specific feature, IMHO.

However, you've somewhat glossed over the other issue - the RuntimeError - and I'm not clear why. To me, it seems a very likely error case - for example:
```
box1 = toga.Box()
widget1 = toga.Button(...)
box1.add(widget1)
box2 = toga.Box()
widget2 = toga.Button(...)
box2.add(widget1)
```

How many buttons are in this layout? There's only 1. In the current setup, you'll get an error on the last line, because you've made a copy-and-paste error; in your version, widget1 will be silently reparented, and it won't be obvious why your second button isn't being displayed.

Deliberate reparenting of a widget is absolutely a use case we need to support - but I don't think it's unreasonable to be explicit about the fact that you are reparenting, especially if it saves you from the accidental reparenting case.

The other part of this PR is the demo app. The demo app seems a lot more complex than it needs to be - why does it have an image, and a secondary button to display a dialog? Unless I'm missing something, all we need is 2 panels of content, each with 1 button. If there's a need to test for more complex view hierarchies, a couple of boxes and other no-op buttons should be sufficient. Any added complexity will only serve to confuse what the app is trying to verify.

In fact - my preference wouldn't be a new test app at all, but to add this functionality test to an existing app (layout and window would be obvious candidates). I realize I wasn't entirely clear about this in my original response, so apologies for the misdirection. For some background - Having dozens of single-purpose test apps makes the release testing process a lot more difficult than it needs to be; I suspect in the very near future we will be reducing the test apps to 2 - one that is a feature demonstrator, and one that is an exhaustive test bed.

File: examples/one_window/one_window/main_box.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import Callable
+
+from travertino.constants import COLUMN
```
[the remark this answers is not in the document]
Comment (saroad2) [reply examples/one_window/one_window/main_box.py:3]: I can combine back everything into *app.py* in a future commit.

File: examples/numberinput/pyproject.toml [hunk 2]
```diff
@@ -2,47 +2,47 @@
 requires = ["briefcase"]
 
 [tool.briefcase]
-project_name = "Demo NumberInput"
+project_name = "Demo One Window"
```
[the remark this answers is not in the document]
Comment (saroad2) [reply examples/numberinput/pyproject.toml:5]: My bad! I accidently switched them.
Will fix that.

**Comment by saroad2** [comment]
Hey @freakboy3742 

From your answer, I believe the solution you want is actually 3.

When replacing the content of the window, even though `box2` is now displayed, the window is still the container if `box1` since no one changed that!
If we keep raising `RuntimeError`, when I'll try to set back `box1` as the content, the `RuntimeError` will be raised since `box1` already have a container, the same window!

Therefore, if we want to keep the `RuntimeError` and avoid implicit reparenting, when witching window contents we need to set **explicitly** the container of the old content to `None` on switch.

What do you think?

**Comment by freakboy3742** [comment]
> Hey @freakboy3742
> 
> From your answer, I believe the solution you want is actually 3.
> 
> When replacing the content of the window, even though `box2` is now displayed, the window is still the container if `box1` since no one changed that! If we keep raising `RuntimeError`, when I'll try to set back `box1` as the content, the `RuntimeError` will be raised since `box1` already have a container, the same window!
> 
> Therefore, if we want to keep the `RuntimeError` and avoid implicit reparenting, when switching window contents, we need to set **explicitly** the container of the old content to `None` on switch.

My apologies - I misinterpreted what you were suggesting with Option 3. On a re-read, I think you're right; this is something that should be reset when you change the window's content. Changing a window's content is an implicit removal of the old content; being explicit about that (and explicitly un-setting container and window associations) sounds like the right approach.

**Comment by mhsmith** [comment]
> Having dozens of single-purpose test apps makes the release testing process a lot more difficult than it needs to be; I suspect in the very near future we will be reducing the test apps to 2 - one that is a feature demonstrator, and one that is an exhaustive test bed.

Strongly agreed: with an app that contains all the widgets, , I could be a lot more confident that that Introducing Chaquopy won't break some Android widget in some way. Even once each Toga platform has good automated testing, there will still be many aspects of the UI that can only be checked interactively.

**Comment by saroad2** [comment]
So I revert the previous fix and wrote a new fix based on approach 3. It required to introduce a new method called `Window.before_set_content`. In that method, I remove the container of the previous widget.

I added an empty implementation of this method in all platforms.

As for the new example, I tried to make an example that replicates the exact use case of mine (and therefore did some copy-paste). I agree that it is too complex for demonstrating the new ability. I'll try to use this new feature in one of the pre-existing examples. Just wanted to know that the implementation is sold first.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Again, this *works*... but the fact that every platform has the concept of a container, but only Windows apparently needs an implementation of `before_set_content` is a red flag for me.

I can't think of anything that is fundamentally *different* about winforms in this regard. This feels like the Winforms backend is missing a critical single line of code clearing the container when widgets are removed, or in cascading that decision to children. What *exactly* is going on here that is causing the problem? Or, what is fundamentally different about Winforms that requires this specific entry point?

**Comment by saroad2** [comment]
When the method `Window.set_content` is called in the implementation level **the content of the window as already switched**. That means that within this method I do not have access to the old content.

Therefore, I had 2 options:
1. Set the container of the old widget in a new method of `Window` that I called `before_set_content`
2. Add this logic to the setter of `Widget.window`, which means that the entire logic will live in the `Widget` class and not the `Window` class.

Option 2 seems like a **big mistake** to me, even though it does not introduce a new function. However, if you really oppose adding new methods to the `Window` implementation, I can go with 2.

**Comment by freakboy3742** [comment]
> When the method `Window.set_content` is called in the implementation level **the content of the window has already switched**. That means that within this method I do not have access to the old content.

Ok - but that doesn't answer the question of why Windows is *different*. Why is this *not* needed on macOS or GTK?

**Comment by saroad2** [comment]
> Ok - but that doesn't answer the question of why Windows is different. Why is this not needed on macOS or GTK?

I think if you'll try to run the "one window" example on Mac it will also won't work, the same way Windows didn't work before this fix.

**Comment by freakboy3742** [comment]
> > Ok - but that doesn't answer the question of why Windows is different. Why is this not needed on macOS or GTK?
> 
> I think if you'll try to run the "one window" example on Mac it will also won't work, the same way Windows didn't work before this fix.

On macOS, it *does* work, which is why I was asking. However, I just noticed that while it is *functionally* working, it is also raising a RuntimeError; I'm not sure why that isn't causing larger problems with the app. GTK is definitely broken, though - the content disappears, but there's no error. I assumed it was working, so apologies for that confusion.

However, on macOS - if you add `child._impl.container = None` before line L197 (i.e., set container to None prior to setting it to widget), the error goes away, and the behavior is as expected.

And... the same is true on Windows. Add `child._impl.container = None` before line L136, and you don't need the `before_set_content` entry point. 

The GTK problem seems to be something else; the same fix would seem to make sense in context, but GTK seems to have an additional problem with widgets not automatically being visible when they're added to a view hierarchy. A naive fix is to call `window.show_all()` after adding the content, but that causes a segfault when you switch content back.

**Comment by saroad2** [comment]
> On macOS, it does work, which is why I was asking. However, I just noticed that while it is functionally working, it is also raising a RuntimeError

I don't have a Mac computer to see it in action, but as I suspected, the `RuntimeError` is raised. The only different is that on Windows it causes crash and in Mac it doesn't.

> And... the same is true on Windows. Add child._impl.container = None before line L136, and you don't need the before_set_content entry point.

While this solution does work, it is not solution 3, that's **solution 2**! I could implement this solution, but pay attention that we do not remove the container of the old widget on switch, we do it when we **set the widget back as content**. This is a whole different solution.

As I said, In my honest opinion, I think my solution here is the best one and can be used in other platform, including Mac which will prevent the `RuntimeError` from being raised.

**Comment by saroad2** [comment]
I removed the "one_window" and added the content replacement into the "window" example.

**Comment by freakboy3742** [comment]
> > On macOS, it does work, which is why I was asking. However, I just noticed that while it is functionally working, it is also raising a RuntimeError
> 
> I don't have a Mac computer to see it in action, but as I suspected, the `RuntimeError` is raised. The only different is that on Windows it causes crash and in Mac it doesn't.
> 
> > And... the same is true on Windows. Add child._impl.container = None before line L136, and you don't need the before_set_content entry point.
> 
> While this solution does work, it is not solution 3, that's **solution 2**! I could implement this solution, but pay attention that we do not remove the container of the old widget on switch, we do it when we **set the widget back as content**. This is a whole different solution.

Ah - now I see it; thanks for the clarification. This general approach makes sense; however, I'd suggest a couple of small tweaks:

1. I'd be inclined to use a name like `clear_content()` rather than `before_set_content()`; the functionality we need to before "before" is very specific, so we might as well communicate that in the function name.
2. In a similar way that the container isn't currently being cleared prior to setting the new container, the old widget is still connected to the original *window*; there should be a call to `self._content.window = None` prior to calling `clear_content()`
3. The current implementation of `before_set_content()` clears the container of the base widget, whereas `set_content()` only assign the container on children. The content widget itself shouldn't have a container, so we only need clear the container on the children. 
4. The implementation of `clear_content()` should be the same on every platform, so you should be able to use a copy of the winforms implementation on every platform, rather than using a no-op. 

Point (4) is probably an indication that there is room for a larger refactor that promotes "container" to an interface level concept. However, that's not a problem we have to solve immediately; it's something worth keeping in mind as we look at top level navigation widgets.

**Comment by saroad2** [comment]
Hey @freakboy3742 .

I did 1-3 but not 4 because I don't have access to other platforms such as Mac and GTK and I could not test it.

**Comment by mhsmith** [comment]
> The implementation of clear_content() should be the same on every platform, so you should be able to use a copy of the winforms implementation on every platform, rather than using a no-op.

There's too much 5x duplicated code in this area already, so if multiple classes have identical implementations of a method, then it's better to have a single implementation in a utility function which they all call. 

Longer-term, if multiple classes have identical implementations of *multiple* methods, then they should probably all inherit from a common base class. And adding abstract functions to the base class would also provide a useful specification of exactly what methods the implementation classes like Widget and Window are required to implement, which is not currently easy to discover.

**Comment by freakboy3742** [comment]
@mhsmith I completely agree that the duplication isn't great; however, it's not quite as simple as factoring the code out into a utility method, because the code being duplicated is in different backends. I guess we *could* put that utility method in toga-core; but at present, we don't use toga-core as a source of utility functionality shared between backends.

However, there's almost certainly an opportunity for refactoring that would move the bulk of the duplicated logic out of the platform backends, and into the toga-core layer. This is what I was alluding to in point (4) above - if containers become an interface-level concept, then the duplicated logic around container management can be centralised in toga-core. This removes the need for backends to access utility functionality at all by altering the backend contract API to be a lower level "how do I respond to being added/removed to a container" response - which is much more in line with the relationship we're trying to keep between interface and implementation.

**Review by freakboy3742 — APPROVED** [review]
Thanks for the work on this. I've added the implementations for GTK, macOS, Android, iOS and Web; the GTK and iOS implementations don't currently work, but the fix in this PR seems generally necessary, and the problem that is occurring doesn't appear to be directly related in both cases.

**Pull request merged**